### PR TITLE
Feature/add findex for redis

### LINKED ISSUE #21: Fix tslint warnings

### PR DESCRIPTION
**Pull request by Manuthor**
---
### Added
- Indexes compaction for Findex
- Also package for CommonJS/Node
- Findex implementation with Redis database (upsert and search)
- Add tests for Cloudproof or Findex only (for PostgRest and Redis)
### Changed
- Reorganize code tree
- Use and apply eslint standard style
### Fixed
- Findex interface with `Uint8Array` in place of `string`
### Removed

Closes #21 (use eslint)

### CODE REVIEW

File: src/crypto/abe/interfaces/encryption_parameters.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  public set additionalData (value: Uint8Array | undefined) {
+    this._additionalData = value
+  }
+
+  constructor (uid: Uint8Array, additionalData?: Uint8Array) {
+    this._uid = uid
+    this._additionalData = additionalData
+  }
+
+  public toJsonEncoded (): Uint8Array {
```
Comment (tbrezot) [inline src/crypto/abe/interfaces/encryption_parameters.ts:39]: Missing documentation on a public function

File: src/crypto/abe/interfaces/encryption_parameters.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    const metadata: any = {}
+    metadata.uid = Array.from(this._uid)
+    if (this._additionalData !== undefined) {
+      metadata.additionalData = Array.from(this._additionalData)
+    }
+    return new TextEncoder().encode(JSON.stringify(metadata))
+  }
+}
+
+// tslint:disable-next-line: max-classes-per-file
```
Comment (tbrezot) [inline src/crypto/abe/interfaces/encryption_parameters.ts:49]: Is this really necessary ? Maybe we can remove it in the next MR (and create a new file for the Metadata class)

File: src/crypto/abe/interfaces/keygen.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+/* tslint:disable:max-classes-per-file */
```
Comment (tbrezot) [inline src/crypto/abe/interfaces/keygen.ts:1]: Is an exported interface counted as a class ? If not maybe you can remove this

File: src/crypto/abe/interfaces/policy.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  private readonly _attributeToInt?: {}
+
+  constructor (axis: PolicyAxis[], maxAttributeCreations: number, lastAttributeValue?: number, attributeToInt?: {}) {
+    this._axis = axis
+    this._maxAttributeCreations = maxAttributeCreations
+    this._lastAttributeValue = lastAttributeValue
+    this._attributeToInt = attributeToInt
+  }
+
+  // Convert this `Policy` to JSON. Output example:
```
Comment (tbrezot) [inline src/crypto/abe/interfaces/policy.ts:55]: Dead code

File: src/crypto/abe/interfaces/policy.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  //         ],
+  //         "Department::R&D": [
+  //         6
+  //         ],
+  //         "Security Level::High Secret": [
+  //         4
+  //         ]
+  //     }
+  // }
+  public toJsonEncoded (): Uint8Array {
```
Comment (tbrezot) [inline src/crypto/abe/interfaces/policy.ts:111]: Missing documentation

File: src/crypto/abe/interfaces/policy.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    const policy: any = {}
+    policy.axes = {}
+    policy.attribute_to_int = {}
+    if (this._lastAttributeValue === undefined && this._lastAttributeValue === undefined) {
```
Comment (tbrezot) [inline src/crypto/abe/interfaces/policy.ts:115]: I think we should rather do this in Rust: we should have a wasm interface in the `abe_policy` crate to convert a policy into Json. It is too easy to introduce incompatibilities if two crates have their own implementation of the serialization.

File: src/crypto/abe/interfaces/policy.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    return result
+  }
+
+  public static fromJsonEncoded (policy: string): Policy {
+    logger.log(() => 'policy: ' + policy)
+    const policyJson = JSON.parse(policy)
+
+    // Fill Policy Axis
+    const axis: PolicyAxis[] = []
+    for (const e of Object.keys(policyJson.axes)) {
```
Comment (tbrezot) [inline src/crypto/abe/interfaces/policy.ts:145]: Same here. Maybe you can create issues about this.

File: src/crypto/hkdf.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import { logger } from 'utils/logger'
+
+/**
+ * HKDF HMAC-SHA256
+ *
+ * @param key hmac key
+ * @param salt salt
+ * @returns a 32 bytes value being the output HMAC
+ */
+export async function hmacSha256 (key: Uint8Array, salt: Uint8Array): Promise<Uint8Array> {
```
Comment (tbrezot) [inline src/crypto/hkdf.ts:11]: Are we really using a JS implementation on HKDF?

File: src/kms/data_structures/KeyBlock.ts [hunk 9]
```diff
@@ -1,121 +1,129 @@
...
-        return "{" + " key_format_type='" + this.key_format_type + "'" + ", key_compression_type='"
-            + this.key_compression_type + "'" + ", key_value='" + this.key_value + "'"
-            + ", cryptographic_algorithm='" + this.cryptographic_algorithm + "'" + ", cryptographic_length='"
-            + this.cryptographic_length + "'" + ", key_wrapping_data='" + this.key_wrapping_data + "'" + "}"
+  public equals (o: any): boolean {
+    if (o == this) { return true }
+    if (!(o instanceof KeyBlock)) {
+      return false
     }
+    const keyBlock = o
```
Comment (tbrezot) [inline src/kms/data_structures/KeyBlock.ts:114]: Why do er need to copy here ?

File: src/kms/decorators/function.ts [hunk 10]
```diff
@@ -1,30 +1,31 @@
...
 
-export function PropertyMetadata(updates: PropertyMetadata) {
-    return (target: any, propertyKey: string | symbol) => {
-        // Pull the existing metadata or create an empty object
-        const allMetadata = (
-            Reflect.getMetadata(METADATA_KEY, target)
-            ||
+/**
+ *
+ * @param updates
```
Comment (tbrezot) [inline src/kms/decorators/function.ts:6]: This documentation does not help me to understand the function :p

File: src/interface/findex/findex.ts [hunk 11]
```diff
@@ -1,87 +1,134 @@
...
+
+    /**
+     * Reset all indexes and upsert new ones
+     *
+     * @param location location string naming the key of location to index
+     */
+    async upsertUsersIndexes(masterKeysFindex: MasterKeys, users: Users, location: string) {
+        // const startDate = new Date().getTime()
+        const generatedUsers = await users.getUsers();
+        // let endDate = new Date().getTime()
```
Comment (tbrezot) [inline src/interface/findex/findex.ts]: dead code

File: src/crypto/sse/findex/interfaces/findex.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+  public get db (): DBInterface {
+    return this._db
+  }
+
+  constructor (db: DBInterface) {
+    this._db = db
+  }
+
+  public async upsert (masterKeys: MasterKeys, label: Uint8Array, locationAndWords: { [key: string]: string[] }): Promise<any> {
```
Comment (tbrezot) [inline src/crypto/sse/findex/interfaces/findex.ts:22]: Public functions of this module do not have documentation

File: src/crypto/abe/core/hybrid_crypto/worker/worker.ts [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+  decrypt (encryptedEntries: Array<{ ciphertextHex: string }>): Uint8Array[] {
+    let dec: HybridDecryption
+    if (this.hybridDecryption === null) {
+      // TODO handle hybrid crypto not initialized here if needed
+      throw new Error('The hybrid decryption scheme is not initialized')
+    } else {
+      dec = this.hybridDecryption
+    }
+
+    const cleartextValues: Uint8Array[] = []
```
[the remark this answers is not in the document]
Comment (Manuthor) [reply src/crypto/abe/core/hybrid_crypto/worker/worker.ts:40]: I am not sure we can initialize an empty array with a given capacity

File: src/crypto/abe/core/hybrid_crypto/worker/worker_pool.ts [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+ *
+ * @param workers
+ * @param decryptionKeyHex
+ * @param encryptedEntries
+ * @param isGpswImplementation
+ * @returns an array of the results that could be decrypted
+ */
+export async function runWorkers (workers: Worker[], decryptionKeyHex: string, encryptedEntries: EncryptedEntry[], isGpswImplementation: boolean): Promise<Uint8Array[]> {
+  logger.log(() => 'NUM WORKERS: ' + workers.length)
+  if (workers.length === 1) {
```
[the remark this answers is not in the document]
Comment (Manuthor) [reply src/crypto/abe/core/hybrid_crypto/worker/worker_pool.ts:58]: Adding a check and throwing an error if that's the case

File: src/crypto/abe/core/keygen/cover_crypt.ts [hunk 15]
```diff
@@ -1,48 +1,47 @@
...
 
 export class CoverCryptMasterKey extends AbeMasterKey {
 
 }
 
-export class CoverCryptMasterKeyGeneration extends AbeKeyGeneration {
-  public generateMasterKey(policy: Policy): AbeMasterKey {
-    logger.log(() => "policy: " + policy)
+export class CoverCryptMasterKeyGeneration implements AbeKeyGeneration {
+  public generateMasterKey (policy: Policy): AbeMasterKey {
```
[the remark this answers is not in the document]
Comment (Manuthor) [reply src/crypto/abe/core/keygen/cover_crypt.ts:13]: I suppose you are refering to the space after the name function: it is done by eslint

File: src/crypto/abe/core/keygen/gpsw_crypt.ts [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+}
+
+export class GpswMasterKeyGeneration implements AbeKeyGeneration {
+  public generateMasterKey (policy: Policy): AbeMasterKey {
+    logger.log(() => `policy: ${policy.toJsonEncoded().toString()}`)
+
+    const policyBytes = policy.toJsonEncoded()
+    const masterKeys = webassembly_generate_master_keys(policyBytes)
+    const privateKeySize = fromBeBytes(masterKeys.slice(0, 4))
```
[the remark this answers is not in the document]
Comment (Manuthor) [reply src/crypto/abe/core/keygen/gpsw_crypt.ts:22]: Yes I changed this in GPSW

File: src/crypto/hkdf.ts (hunk 8 above)
Comment (Manuthor) [reply src/crypto/hkdf.ts:11]: deleting

File: src/crypto/abe/core/hybrid_crypto/worker/worker.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (tbrezot) [reply src/crypto/abe/core/hybrid_crypto/worker/worker.ts]: I think this works:
``` ts
new Uint8Array(length)
```

**Pull request merged**